### 데이터 가져오기
---

![image1](https://aiffelstaticprd.blob.core.windows.net/media/images/F25-2.max-800x600.png)

### 소켓통신

In [1]:
import socket
with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:     # 소켓 s를 하나 열어서
    s.connect(('www.google.com', 80))          # www.google.com 이라는 호스트의 80번 포트에 접속합니다. 
    cmd = 'GET http://www.google.com HTTP/1.0\r\n\r\n'.encode() 
    s.send(cmd)                 # 접속이 되면 GET xxxxx 형식의 HTTP 프로토콜로 된 명령을 전송합니다. 
    data = s.recv(1024)    # 그리고 1024바이트만큼의 HTTP 응답을 수신합니다. 
print('Received:', data)  

Received: b'HTTP/1.0 404 Not Found\r\nContent-Type: text/html; charset=UTF-8\r\nDate: Mon, 15 Feb 2021 00:40:55 GMT\r\nServer: gws\r\nContent-Length: 1561\r\nX-XSS-Protection: 0\r\nX-Frame-Options: SAMEORIGIN\r\n\r\n<!DOCTYPE html>\n<html lang=en>\n  <meta charset=utf-8>\n  <meta name=viewport content="initial-scale=1, minimum-scale=1, width=device-width">\n  <title>Error 404 (Not Found)!!1</title>\n  <style>\n    *{margin:0;padding:0}html,code{font:15px/22px arial,sans-serif}html{background:#fff;color:#222;padding:15px}body{margin:7% auto 0;max-width:390px;min-height:180px;padding:30px 0 15px}* > body{background:url(//www.google.com/images/errors/robot.png) 100% 5px no-repeat;padding-right:205px}p{margin:11px 0 22px;overflow:hidden}ins{color:#777;text-decoration:none}a img{border:0}@media screen and (max-width:772px){body{background:none;margin-top:0;max-width:none;padding-right:0}}#logo{background:url(//www.google.com/images/branding/googlelogo/1x/googlelogo_color_150x54dp.png) no-re

### TCP/IP

1. TCP: Transmission Control Protocol -> 소켓 포트 단위의 송수신 프로토콜
    * Transport 레이어에서 어플리케이션 간의 통신 채널을 생성하고 패킷으로 쪼개진 데이터의 정확한 전송 흐름을 관리한다. 
    * 하나의 컴퓨터에 여러개의 어플리케이션이 구동되고 있다면 각각 다른 포트 번호의 소켓을 가지고 있게 되는데, 패킷 데이터가 누락되거나 순서가 바뀌는       일 없이, 이 포트 번호를 기준으로 타겟 어플리케이션 소켓으로 정확히 전송될 수 있도록 데이터 흐름을 책임지는 것이 TCP의 역할이다.   
2. IP: Internet Protocol -> 컴퓨터마다 주어지는 IP 주소 단위의 송수신 프로토콜
   * IP는 패킷이 전달되어야 할 주소(IP주소)를 정의한다.
   * 주소 체계에 따라 목적지로 패킷들을 전송하기 위해 거쳐야 할 다음 라우팅 경로를 결정할 수 있게 한다.
   * 패킷의 목적지가 되는 컴퓨터(호스트)는 저마다의 IP 주소를 할당받는다.
   * IP는 타겟 주소로 데이터가 정확히 전송되는 것을 책임지는 역할을 하는 프로토콜이다.
     
3. TCP/IP 기반의 Application 프로토콜
   * HTTP: HyperText Transfer Protocol
   * HTTPS: HyperText Transfer Protocol Secure
   * FTP: File Transfer Protocol
   * SSH: Secure Shell (암호화된 네트워크 프로토콜)
   * DNS: Domain Name System
  
4. Port Number
   ![image](https://aiffelstaticprd.blob.core.windows.net/media/original_images/F25-8.png)

### API
![api](https://aiffelstaticprd.blob.core.windows.net/media/images/F25-13.max-800x600.png)

### OPEN API

1. [구글 API](https://developers.google.com/apis-explorer)
2. [카카오 API](https://developers.kakao.com/)
3. [트위터 API](https://developer.twitter.com/en/docs/api-reference-index)
4. [네이버 파파고 API](https://developers.naver.com/docs/papago/)

### Rest API
#### 1. REST란?   
    Representational State Transfe라는 용어의 약자이다.   
    자원을 URI로 표시하고 해당 자원의 상태를 주고 받는 것을 의미한다.
   
#### 2. REST의 구성 요소   
    자원(Resource): URI   
    행위(Verb): HTTP METHOD   
    표현(Representations)   

#### 3. REST의 특징   
   1. Uniform Interface (유니폼 인터페이스)
        HTTP 표준만 따른다면 어떤 언어 혹은 어떤 플랫폼에서 사용하여도 사용이 가능한 인터페이스 스타일이다.
        안드로이드 플랫폼, IOS 플랫폼 등 특정 언어나 플랫폼에 종속되지 않고 사용이 가능하다.
   2. Stateless (상태 정보 유지 안함)
        Rest는 상태 정보를 유지하지 않는다.
        서버는 각각의 요청을 완전히 다른 것으로 인식하고 처리를 한다.
        이전 요청이 다음 요청 처리에 연관이 되면 안된다.
   3. Cacheable (캐시 가능)
        HTTP의 기존 웹 표준을 그대로 사용하기 때문에 HTTP가 가진 캐싱 기능 적용이 가능하다.
   4. Self-descriptiveness (자체 표현 구조)
        Rest API 메시지만 보고도 쉽게 이해할 수 있는 자체 표현 구조로 되어있다.
   5. Client-Server
        Rest 서버는 API 제공을 하고 클라이언트는 사용자 인증에 관련된 일들을 직접 관리한다.
        자원이 있는 쪽을 Server라고 하고 자원을 요청하는 쪽이 Client가 된다.
        서로간의 의존성이 줄어들기 때문에 역할이 확실하게 구분되어 개발해야할 내용들이 명확해진다.
   6. Layerd System (계층화)
        클라이언트는 Rest API 서버만 호출한다.
        Rest 서버는 다중 계층으로 구성될수 있으면 로드 밸런싱, 암호화, 사용자 인증 등을 추가하여 구조상의
        유연성을 둘 수 있다.

#### 4. REST API란?
    Rest 기반의 규칙들을 지켜서 설계된 API를 Rest API 혹은 Restful API이라고 한다.

#### 5. REST API 설계 규칙
    1. URI는 정보의 자원을 표현해야한다.
        자원의 이름은 동사보다는 명사를 사용한다.
        URI는 자원을 표현하는데 중점을 두어야 하기 때문에 행위에 대한 표현이 들어가면 안된다.
        (URI에 HTTP METHOD와 행위에대한 동사 표현이 들어가면 안된다.)
        
        GET /users/321   
        
    2. 자원에 대한 행위는 HTTP METHOD로 표현한다. (GET, POST, PUT DELETE)
        URI에 자원의 행위에 대한 표현이 들어가지 않는 대신 HTTP METHOD를 통해 대신한다.
    
        GET /users/321 321 ID를 가진 유저 정보를 가져오기
        DELETE /users/321 321 ID를 가진 유저 정보를 삭제하기
        POST /users 유저를 생성하기
    
    3. 슬래시 (/)는 계층 관계를 나타내는데 사용한다.

        http://restapi.test.com/users/rooms
        http://restapi.test.com/users/board

    4. URI 마지막은 슬래시(/)를 사용하면 안된다.

        http://restapi.test.com/users/rooms/ [X]
        http://restapi.test.com/users/rooms  [O]

    5. 하이픈 (-)은 URI의 가독성을 높이는데 사용한다.
        불가피하게 긴 URI를 사용하게 될 경우 하이픈을 이용하여 가독성을 높인다.

    6. 언더바(_) 혹은 밑줄은 URI에 사용하지 않는다.
        밑줄은 보기 어렵거나 밑줄 때문에 문자가 가려지기도 한다.
        그래서 대신 언더바를 사용하지 않고 하이픈을 이용한다.

    7. URI는 경로에는 소문자가 적합하다.
        URI 경로에는 대문자 사용을 피해야한다.
        대소문자에 따라 각자 다른 리소스로 인식하기 때문이다.
        RFC3986(URI 문법 형식)은 URI 스키마와 호스트를 제외하고는 대소문자를 구별하도록 규정하기 때문이다.
    
    8. 파일 확장자는 URI에 포함하지 않는다.
        REST API에서는 메시지 바디 내용의 포맷을 나타내기 위한 파일 확장자를 URI 안에 포함시키지 않는다.
        Accept header를 사용한다.
    
    9. 리소스 간의 관계를 표현하는 방법

        GET : /users/{userid}/devices   
    
#### 6. HTTP 응답 코드
* 2xx 성공<br>
    200: 클라이언트의 요청을 정상적으로 수행함.<br>
    201: 클라이언트에게 생성 작업을 요청 받았고, 생성 작업을 성공함.<br>
    204: 요청은 성공 했지만 응답할 콘텐츠가 없음.<br>
* 3xx 리다이렉션<br>
    301: 클라이언트가 요청한 리소스에 대한 URI가 영구적으로 변경되었을 때 사용함.<br>
    302: 301과 같으나 임시적으로 주소가 바뀌었을 경우 사용함.<br>
    304: 이전에 방문했을 때의 요청 결과와 다르지 않을 경우 사용함. 캐시된 페이지를 그대로 사용.<br>
    307: 임시 페이지로 리다이렉트.<br>
* 4xx 클라이언트 오류<br>
    400: 클라이언트가 올바르지 못한 요청을 보냄.<br>
    401: 로그인을 하지 않아 페이지를 열 권한이 없음.<br>
    403: 금지된 페이지, 로그인을 하든 안하든 접근할 수 없음. (관리자 페이지)<br>
    404: 찾을 수 없는 페이지, 주소를 잘 못 입력했을 때 사용함.<br>
    403 대신에 사용할 수도 있음.(해커들의 공격을 방지하고자 페이지가 없는 것처럼 위장함)<br>
    408: 요청 시간이 초과됨.<br>
    409: 서버가 요청을 처리하는 과정에서 충돌이 발생한 경우. (회원가입 중 중복된 아이디인 경우)<br>
    410: 영구적으로 사용할 수 없는 페이지.<br>
* 5xx 서버 오류<br>
    501: 해당 요청을 처리하는 기능이 만들어지지 않음.<br>
    502: 서버로 가능 요청이 중간에서 유실된 경우.<br>
    503: 서버가 터졌거나 유지 보수 중 (유지 보수 중일때는 유지 보수중이라는 것을 알려주는 페이지로 전송해주는 것이 좋음)<br>
    504: 서버 게이트웨이에 문제가 생겨 시간 초과가 된 경우.<br>
    505: HTTP 버전이 달라 요청이 처리할 수 없음.<br>


### 날씨 OPEN API 사용해보기

In [2]:
import requests
myapi = "afd7bfb9b5a9b81537b9cab2435e2a8b"

#lat: 위도, #lon:경도
def makeurl(lat, lng):
    url = "https://api.openweathermap.org/data/2.5/onecall?lat={}&lon={}&appid={}".format(lat, lng, myapi)
    return url

response = requests.get(makeurl(37,127))
weather = response.json()
weather

{'lat': 37,
 'lon': 127,
 'timezone': 'Asia/Seoul',
 'timezone_offset': 32400,
 'current': {'dt': 1613352477,
  'sunrise': 1613341269,
  'sunset': 1613380284,
  'temp': 275.15,
  'feels_like': 268.49,
  'pressure': 1011,
  'humidity': 87,
  'dew_point': 273.22,
  'uvi': 1.34,
  'clouds': 90,
  'visibility': 4828,
  'wind_speed': 6.69,
  'wind_deg': 290,
  'wind_gust': 9.26,
  'weather': [{'id': 804,
    'main': 'Clouds',
    'description': 'overcast clouds',
    'icon': '04d'}]},
 'minutely': [{'dt': 1613352480, 'precipitation': 0},
  {'dt': 1613352540, 'precipitation': 0},
  {'dt': 1613352600, 'precipitation': 0},
  {'dt': 1613352660, 'precipitation': 0},
  {'dt': 1613352720, 'precipitation': 0},
  {'dt': 1613352780, 'precipitation': 0},
  {'dt': 1613352840, 'precipitation': 0},
  {'dt': 1613352900, 'precipitation': 0},
  {'dt': 1613352960, 'precipitation': 0},
  {'dt': 1613353020, 'precipitation': 0},
  {'dt': 1613353080, 'precipitation': 0},
  {'dt': 1613353140, 'precipitation': 0},

### 크롤링 이용해보기

#### 1. 웹페이지 다운로드

In [3]:
# urlopen()함수를 통해 해당 서버에서 요청받은 ("HTTPResponse")객체를 반환 합니다.
import urllib
def download(url):
    return urllib.request.urlopen(url)

In [4]:
from urllib.error import URLError, HTTPError, ContentTooShortError

def download(url):
    try:
        html = urllib.request.urlopen(url)
    except (URLError, HTTPError, ContentTooShortError) as e:
        print('Download error', e.reason)
        html = None
    return html

download('https://www.google.com')

#### 2. 웹페이지 분석하기

In [5]:
# HTML Data 파싱
def download(url):
    try:
        html = urllib.request.urlopen(url).read()
    except (URLError, HTTPError, ContentTooShortError) as e:
        print('Download error', e.reason)
        html = None
    return html

In [6]:
download('https://www.google.com')

b'<!doctype html><html itemscope="" itemtype="http://schema.org/WebPage" lang="ko"><head><meta content="text/html; charset=UTF-8" http-equiv="Content-Type"><meta content="/images/branding/googleg/1x/googleg_standard_color_128dp.png" itemprop="image"><title>Google</title><script nonce="2tCazhCmnL0BrKSzxIjvpA==">(function(){window.google={kEI:\'Ss8pYJvhEIXN-QaJxbeAAg\',kEXPI:\'0,1302433,56976,954,756,4348,207,2415,701,88,10,1590,2316,383,246,5,1354,1300,3951,5,1660,676,2998,7,1117169,1232,1196561,490,328984,51224,16114,28684,9188,8384,4858,1362,9291,3028,4739,7,1841,4720,6273,4020,978,13228,2974,873,37,4155,6430,14528,4516,2778,919,2277,8,2796,889,704,1279,2212,530,149,1103,840,520,1519,4258,312,1137,2,2063,606,2023,1777,520,4166,3,100,328,1284,2943,5846,3227,2845,7,5599,6755,5096,7540,336,4929,108,2854,554,907,2,941,2614,2397,1387,4948,1133,3277,3,346,230,970,865,4624,149,5990,7985,4,1528,2304,1236,1145,4658,1791,1373,1519,460,1555,4067,5634,1426,714,1770,3011,1753,2658,4243,518,912,565

#### 3. 파이썬 라이브러리 BeautifulSoup, Requests 사용하여 쉽게 웹 크롤링하기
* pip install bs4
* pip install requests

In [7]:
# 웹페이지 다운로드
import requests
url = 'http://www.google.com'
response = requests.get(url)
response

<Response [200]>

In [8]:
def download2(url):
    try:
        response = requests.get(url)
        html = response.text
    except requests.ConnectionError:
        print('Connection error')
        html = None
    return html

download2('https://www.google.com')

'<!doctype html><html itemscope="" itemtype="http://schema.org/WebPage" lang="ko"><head><meta content="text/html; charset=UTF-8" http-equiv="Content-Type"><meta content="/images/branding/googleg/1x/googleg_standard_color_128dp.png" itemprop="image"><title>Google</title><script nonce="tG/4BkLneG0cf5R9vnT1rQ==">(function(){window.google={kEI:\'x88pYPXYLKLw-QaCrKv4Ag\',kEXPI:\'0,1359409,954,5104,207,3204,10,1590,2316,383,246,5,1129,225,1301,3949,2342,3179,7,1116988,1232,1196511,539,328985,51223,16115,17444,11240,9188,8384,4859,1361,284,9006,3027,4741,12841,4020,978,13228,2054,920,873,4192,6430,14528,4517,2777,919,2277,8,3681,708,1279,2212,530,149,1103,840,517,1522,158,4100,108,204,1137,2,2063,606,2023,1777,520,1704,2472,93,328,1284,2943,5846,3227,2845,7,4773,826,6755,5096,7877,3748,1180,108,3407,908,2,940,2615,2397,1387,6081,3277,3,576,970,865,2550,2074,149,5992,7983,4,498,950,80,2304,1236,1145,4658,1791,2893,459,1555,4067,1036,1928,2670,1426,714,1579,3202,1753,2658,4242,519,913,563,464,3

In [9]:
# 웹페이지 분석하기 BeautifulSoup
import requests
from bs4 import BeautifulSoup
html = requests.get('http://www.google.com')
soup = BeautifulSoup(html.text, 'html.parser')

In [10]:
soup.html.body

<body bgcolor="#fff"><script nonce="GUWzDLiBQTiCpbcaY1FoxA==">(function(){var src='/images/nav_logo229.png';var iesg=false;document.body.onload = function(){window.n && window.n();if (document.images){new Image().src=src;}
if (!iesg){document.f&&document.f.q.focus();document.gbqf&&document.gbqf.q.focus();}
}
})();</script><div id="mngb"><div id="gbar"><nobr><b class="gb1">검색</b> <a class="gb1" href="http://www.google.co.kr/imghp?hl=ko&amp;tab=wi">이미지</a> <a class="gb1" href="http://maps.google.co.kr/maps?hl=ko&amp;tab=wl">지도</a> <a class="gb1" href="https://play.google.com/?hl=ko&amp;tab=w8">Play</a> <a class="gb1" href="http://www.youtube.com/?gl=KR&amp;tab=w1">YouTube</a> <a class="gb1" href="https://news.google.com/?tab=wn">뉴스</a> <a class="gb1" href="https://mail.google.com/mail/?tab=wm">Gmail</a> <a class="gb1" href="https://drive.google.com/?tab=wo">드라이브</a> <a class="gb1" href="https://www.google.co.kr/intl/ko/about/products?tab=wh" style="text-decoration:none"><u>더보기</u> »</a><

In [11]:
soup = BeautifulSoup("<span>Wow it's so good!!</span>", 'html.parser')
soup.span

<span>Wow it's so good!!</span>

In [12]:
# ''', """ 는 여러줄의 문자열을 입력할 때 사용
html='''<title>Fundamental</title> 
         <body>
          <p id='programming'>python</p> 
          <p id='programming'>java</p> 
          <p id='algorithm'>algorithm</p> 
          <p id='fundamental'>math</p> 
          <p id='programming'>C++</p> 
          </body>'''
soup = BeautifulSoup(html, 'html.parser')
soup.findAll({'p'})

[<p id="programming">python</p>,
 <p id="programming">java</p>,
 <p id="algorithm">algorithm</p>,
 <p id="fundamental">math</p>,
 <p id="programming">C++</p>]

In [13]:
soup.findAll('p', id='programming')

[<p id="programming">python</p>,
 <p id="programming">java</p>,
 <p id="programming">C++</p>]

[Beatiful soup 참고문서](https://www.crummy.com/software/BeautifulSoup/bs4/doc/)

#### 4. 네이버 환율정보 크롤링해보기

In [14]:
from bs4 import BeautifulSoup
import requests

#URL 가져오기
url = 'https://m.stock.naver.com/marketindex/index.nhn'
response = requests.get(url)

#Soup 객체 생성
soup = BeautifulSoup(response.text, 'html.parser')

#원하는 데이터 추출 - 국가
country = []
country_1 = soup.findAll('strong')
country_2 = soup.findAll('span', {'class':'stock_item'})
for c in country_1:
    country.append(c.text)
for c in country_2:
    country.append(c.text)
country.remove('코스콤')
country.remove('에프앤가이드')
country.remove('레피니티브')
country.remove('이데일리')
country.remove('모닝스타')
country.remove('제로인')
country.remove('한국예탁결제원')
country.remove('IR큐더스')
country.remove('지속가능발전소')

#원하는 데이터 추출 - 환율정보
price = []
price_0 = soup.findAll('span', {"class":"stock_price"})
for p in price_0:
    price.append(p.text)

#데이터 정렬, (pandas)
import pandas as pd
data = {}
for i in range(len(country)):
    data[country[i]] = price[i]

pd.Series(data)

미국 USD        1,104.30
유럽 EUR        1,339.18
일본 JPY        1,051.46
중국 CNY          172.01
국제금           1,821.60
유가 WTI           59.47
달러인덱스            90.47
남아프리카 ZAR        75.98
노르웨이 NOK        130.83
뉴질랜드 NZD        797.80
대만 TWD           39.42
덴마크 DKK         180.07
러시아 RUB          14.99
말레이시아 MYR       273.17
멕시코 MXN          55.43
몽골 MNT            0.39
바레인 BHD       2,929.18
방글라데시 BDT        13.02
베트남 VND           4.80
브라질 BRL         205.64
브루나이 BND        833.78
사우디 SAR         294.39
스웨덴 SEK         133.07
스위스 CHF       1,238.63
싱가포르 SGD        833.78
아랍에미리트 AED      300.64
영국 GBP        1,533.71
오만 OMR        2,868.39
요르단 JOD       1,557.55
이스라엘 ILS        339.52
이집트 EGP          70.98
인도 INR           15.21
인도네시아 IDR         7.92
체코 CZK           52.02
칠레 CLP            1.53
카자흐스탄 KZT         2.63
카타르 QAR         300.92
캐나다 CAD         870.90
쿠웨이트 KWD      3,650.34
태국 THB           36.96
터키 TRY          157.15
파키스탄 PKR          6.95
폴란드 PLN         297.89
필리핀 PHP    

#### 5. 로그인하기
![login](https://aiffelstaticprd.blob.core.windows.net/media/images/F25-23.max-800x600.png)

In [15]:
#세션 생성하기
session = requests.session()

#로그인하기
log_info = {'id':'', #여러분의 ID와 PW를 입력해 주세요
            'pw':''}
url = 'https://nid.naver.com/nidlogin.login'
response = session.post(url, data=log_info)

#마이페이지 접근하기
url_mypage = 'https://nid.naver.com/user2/help/myInfo.nhn?lang=ko_KR'
response = session.get(url_mypage)

#Soup객체 생성
soup = BeautifulSoup(response.text, 'html.parser')

#마이페이지에 별명 가져 오기
soup.find('span')

<span>본문으로 바로가기</span>

#### 6. 브라우저 제어를 통한 크롤링 - 셀레니움 써보기
1. 셀레니움 라이브러리 설치
    * pip install selenium 
    * conda install selenium( 위의 방법이 되지않을시 ) 
2. 브라우저 드라이버 설치
    * 크롬의 경우 설정- 정보에서 버전확인 후 https://sites.google.com/a/chromium.org/chromedriver/downloads에서 다운로드
    * mkdir -p ~/aiffel/lib
    * cd ~/Downloads    
    * unzip chromedriver_linux64.zip
    * mv chromedriver ~/aiffel/lib   
    
3. https://data.seoul.go.kr/에서 산과 공원 데이터 크롤링

In [16]:
# URL 다운로드
crawling_urls = {
"산과공원": "https://data.seoul.go.kr/dataList/OA-12962/S/1/datasetView.do",
}

In [17]:
import requests
from selenium import webdriver
import time
import pandas as pd
import os
import os.path
import glob

In [18]:
# find_element_by_css_selector() 를 이용하여 제어
# 개발자도구를 켠 후 Ctrl+Shift+C를 통해 해당 html 검색 
import os

wd_path = os.getenv('HOME')+'/aiffel/lib/chromedriver'

#웹드라이버 실행 및 페이지 이동
driver = webdriver.Chrome(wd_path)   # 크롬드라이버를 통해 브라우저를 띄우고
driver.get(crawling_urls['산과공원'])      # 우리가 원하는 URL로 이동합니다.
time.sleep(5)       # 해당 화면이 다 로딩할 때까지 5초간 충분히 기다려 줍니다. 
    
#csv파일 다운로드 버튼 클릭하기
driver.find_element_by_css_selector("#btnCsv").click()   # 사람이 누른 것처럼 다운로드 버튼을 클릭한 후, id값을 select
time.sleep(3)     # 다운로드가 완료될 때까지 3초간 기다려 줍니다. 

driver.quit()      # 브라우저를 닫습니다.

In [19]:
#다운받은 csv파일 확인
_dir = os.getenv('HOME')+'/Downloads'   # 영문 우분투 사용자라면 이 경로를 선택해 주세요.

files = glob.glob('{}/서울시*.csv'.format(_dir))
print(files)

['/home/ssac24/Downloads/서울시 산과공원 생태관광 정보 (한국어).csv', '/home/ssac24/Downloads/서울시 산과공원 생태관광 정보 (한국어) (1).csv']


In [33]:
#csv파일을 dataframe으로 변환하기
#인코딩 에러 발생시에 encoding옵션 추가
f_M_park = pd.read_csv(files[0], encoding='CP949', engine='python')   #CP949: windows에서 사용하는 인코딩 방식
f_M_park.head(3)

키             명칭   대분류  \
0  BE_IW14-0020  진관내동 생태경관보전지역  생태탐방   
1  BE_IW14-0109           안산공원  산과공원   
2  BE_IW14-0110          여의도공원  산과공원   

                                           주소   행정 시  행정 구  행정 동  \
0  은평구 진관동 282-1번지 일대(북한산국립공원 북한산성 입구 주변 습지 )  서울특별시   은평구   진관동   
1                     서울특별시 서대문구 홍제동 산33번지 일대  서울특별시  서대문구  홍제1동   
2                서울특별시 영등포구 여의공원로68(여의도동 2번지)  서울특별시  영등포구   여의동   

                         대표전화      면적            지정일  \
0  02-2115-7550~5 02-350-1397  16639㎡  2002년 12월 30일   
1                 02-330-1395     NaN            NaN   
2                 02-761-4079     NaN            NaN   

                                                교?안내  
0  지하철 3호선 구파발역 1번 출구에서 704번 34번 버스를 타고 북한산성 입구에서...  
1                                                NaN  
2                                                NaN

#### [공공데이터 csv파일 코딩이 깨질 시 해결 방법](https://teddylee777.github.io/pandas/%EA%B3%B5%EA%B3%B5%EB%8D%B0%EC%9D%B4%ED%84%B0-%ED%95%9C%EA%B8%80%EA%B9%A8%EC%A7%90%ED%98%84%EC%83%81-%ED%95%B4%EA%B2%B0%EB%B0%A9%EB%B2%95)